Error "Warehouse id: 0 not  exist."

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
1. Create a frontend & backend in the same docker compose file.
2. Start docker compose & add the backend as a backend in the front end.
3. Create a database & a table under it.
4. Create a routine load to the table from a kafka stream.
5. Stop both instances by pressing CTRL+C
6. Wait for both all instances to stop
7. Start up the instances again

### Expected behavior (Required)
```
starrocks-fe-0  | WARNING: Exception thrown during asynchronous load
starrocks-fe-0  | java.util.concurrent.CompletionException: com.starrocks.common.ErrorReportException: Warehouse id: 0 not 
exist.
starrocks-fe-0  |       at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)    
starrocks-fe-0  |       at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)  
starrocks-fe-0  |       at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)   
starrocks-fe-0  |       at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)       
starrocks-fe-0  |       at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)       
starrocks-fe-0  |       at java.base/java.lang.Thread.run(Thread.java:829)
starrocks-fe-0  | Caused by: com.starrocks.common.ErrorReportException: Warehouse id: 0 not exist.
starrocks-fe-0  |       at com.starrocks.common.ErrorReportException.report(ErrorReportException.java:38)
starrocks-fe-0  |       at com.starrocks.server.WarehouseManager.getWarehouse(WarehouseManager.java:91)
starrocks-fe-0  |       at com.starrocks.server.WarehouseManager.getBackgroundWarehouse(WarehouseManager.java:258)
starrocks-fe-0  |       at com.starrocks.statistic.StatisticUtils.buildConnectContext(StatisticUtils.java:111)
starrocks-fe-0  |       at com.starrocks.sql.optimizer.statistics.ColumnBasicStatsCacheLoader.lambda$asyncLoadAll$1(ColumnBasicStatsCacheLoader.java:91)
starrocks-fe-0  |       at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)   
starrocks-fe-0  |       ... 3 more
```

### Real behavior (Required)
Nothing.

### StarRocks version (Required)
 - 3.3.3-312ed45

@kevincai I encountered the same issue as well.    
`Nov 24, 2024 4:00:38 PM com.github.benmanes.caffeine.cache.LocalAsyncCache$AsyncBulkCompleter accept
WARNING: Exception thrown during asynchronous load
java.util.concurrent.CompletionException: com.starrocks.common.ErrorReportException: Warehouse id: 0 not exist.
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.starrocks.common.ErrorReportException: Warehouse id: 0 not exist.
        at com.starrocks.common.ErrorReportException.report(ErrorReportException.java:38)
        at com.starrocks.server.WarehouseManager.getWarehouse(WarehouseManager.java:91)
        at com.starrocks.server.WarehouseManager.getBackgroundWarehouse(WarehouseManager.java:258)
        at com.starrocks.statistic.StatisticUtils.buildConnectContext(StatisticUtils.java:113)
        at com.starrocks.sql.optimizer.statistics.TableStatsCacheLoader.lambda$asyncLoadAll$2(TableStatsCacheLoader.java:67)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
        ... 3 more

Nov 24, 2024 4:00:38 PM com.github.benmanes.caffeine.cache.LocalAsyncCache$AsyncBulkCompleter accept
WARNING: Exception thrown during asynchronous load
java.util.concurrent.CompletionException: com.starrocks.common.ErrorReportException: Warehouse id: 0 not exist.
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.starrocks.common.ErrorReportException: Warehouse id: 0 not exist.
        at com.starrocks.common.ErrorReportException.report(ErrorReportException.java:38)
        at com.starrocks.server.WarehouseManager.getWarehouse(WarehouseManager.java:91)
        at com.starrocks.server.WarehouseManager.getBackgroundWarehouse(WarehouseManager.java:258)
        at com.starrocks.statistic.StatisticUtils.buildConnectContext(StatisticUtils.java:113)
        at com.starrocks.sql.optimizer.statistics.ColumnBasicStatsCacheLoader.lambda$asyncLoadAll$1(ColumnBasicStatsCacheLoader.java:91)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
        ... 3 more

Nov 24, 2024 4:00:38 PM com.github.benmanes.caffeine.cache.LocalAsyncCache$AsyncBulkCompleter accept
WARNING: Exception thrown during asynchronous load
java.util.concurrent.CompletionException: com.starrocks.common.ErrorReportException: Warehouse id: 0 not exist.
        at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
        at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1702)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: com.starrocks.common.ErrorReportException: Warehouse id: 0 not exist.
        at com.starrocks.common.ErrorReportException.report(ErrorReportException.java:38)
        at com.starrocks.server.WarehouseManager.getWarehouse(WarehouseManager.java:91)
        at com.starrocks.server.WarehouseManager.getBackgroundWarehouse(WarehouseManager.java:258)
        at com.starrocks.statistic.StatisticUtils.buildConnectContext(StatisticUtils.java:113)
        at com.starrocks.sql.optimizer.statistics.TableStatsCacheLoader.lambda$asyncLoadAll$2(TableStatsCacheLoader.java:67)
        at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1700)
        ... 3 more`

--- a/fe/fe-core/src/main/java/com/starrocks/journal/GlobalStateCheckpointWorker.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/GlobalStateCheckpointWorker.java
@@ -36,6 +36,8 @@ public class GlobalStateCheckpointWorker extends CheckpointWorker {
         globalStateMgr.setJournal(journal);
         try {
             globalStateMgr.loadImage();
+            // This is required because the loadImage() may find no image to load, and the initDefaultWarehouse() will
+            // be skipped.
             globalStateMgr.initDefaultWarehouse();
 
             checkEpoch(epoch);

--- a/fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java
+++ b/fe/fe-core/src/main/java/com/starrocks/server/GlobalStateMgr.java
@@ -1271,6 +1271,11 @@ public class GlobalStateMgr {
             replayer = null;
         }
 
+        if (!isDefaultWarehouseCreated) {
+            // A brand-new cluster was up for the first time, the leader node initializes its default warehouse here.
+            initDefaultWarehouse();
+        }
+
         // set this after replay thread stopped. to avoid replay thread modify them.
         isReady.set(false);
 
@@ -1311,10 +1316,6 @@ public class GlobalStateMgr {
             // start other daemon threads that should run on all FEs
             startAllNodeTypeDaemonThreads();
             insertOverwriteJobMgr.cancelRunningJobs();
-
-            if (!isDefaultWarehouseCreated) {
-                initDefaultWarehouse();
-            }
 
             MetricRepo.init();
 
@@ -1524,18 +1525,18 @@ public class GlobalStateMgr {
             return;
         }
 
-        // transfer from INIT/UNKNOWN to OBSERVER/FOLLOWER
+        if (!isDefaultWarehouseCreated) {
+            // A brand-new cluster was up for the first time, the follower/observer node initializes its default warehouse here.
+            initDefaultWarehouse();
+        }
 
+        // transfer from INIT/UNKNOWN to OBSERVER/FOLLOWER
         if (replayer == null) {
             createReplayer();
             replayer.start();
         }
 
         startAllNodeTypeDaemonThreads();
-
-        if (!isDefaultWarehouseCreated) {
-            initDefaultWarehouse();
-        }
 
         MetricRepo.init();
 

--- a/fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java
+++ b/fe/fe-core/src/main/java/com/starrocks/server/WarehouseManager.java
@@ -87,10 +87,14 @@ public class WarehouseManager implements Writable {
 
     public void initDefaultWarehouse() {
         // gen a default warehouse
+        // Be sure to make this initialization idempotent, the upper level may invoke it multiple times without
+        // knowing it is initialized or not.
         try (LockCloseable ignored = new LockCloseable(rwLock.writeLock())) {
-            Warehouse wh = new DefaultWarehouse(DEFAULT_WAREHOUSE_ID, DEFAULT_WAREHOUSE_NAME);
-            nameToWh.put(wh.getName(), wh);
-            idToWh.put(wh.getId(), wh);
+            if (!nameToWh.containsKey(DEFAULT_WAREHOUSE_NAME)) {
+                Warehouse wh = new DefaultWarehouse(DEFAULT_WAREHOUSE_ID, DEFAULT_WAREHOUSE_NAME);
+                nameToWh.put(wh.getName(), wh);
+                idToWh.put(wh.getId(), wh);
+            }
         }
     }
 
@@ -379,6 +383,11 @@ public class WarehouseManager implements Writable {
 
     public void load(SRMetaBlockReader reader)
             throws SRMetaBlockEOFException, IOException, SRMetaBlockException {
+        // Create the default warehouse during the WarehouseManager::load() invoked by loadImage()
+        // The default_warehouse is not persisted through WarehouseManager::save(), but some of the image loads and
+        // postImageLoad actions may depend on default_warehouse to perform actions.
+        // The default_warehouse must be ready before postImageLoad.
+        initDefaultWarehouse();
     }
 
     public void addWarehouse(Warehouse warehouse) {

